File: day10/Day10.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Day10 {

    static int[][] grid;
    static int rows, cols;

    //directions
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static void main(String[] args) {
        Path filepath = Path.of("day10_inputs.txt");

        try {
        List<String> lines = Files.readAllLines(filepath);
        // String[] data = lines.toArray(String[]::new);
        rows = lines.size();
        cols = lines.get(0).length();
        grid = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                grid[i][j] = lines.get(i).charAt(j) - '0';
            }
        }

        System.out.println("Part 1: " + SolvePart1());
        System.out.println("Part 2: " + SolvePart2());

        } catch (IOException e) {
            System.err.println("Error: opening the file " + e.getMessage());
        }
    }

    static int SolvePart1() {
        int total = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 0) {
                    Set<String> reachedNines = new HashSet<>();
                    dfsPart1(i, j, reachedNines);
                    // System.out.println(reachedNines);
                    total += reachedNines.size();
                }
            }
        }
        return total;
    }

    static void dfsPart1(int r, int c, Set<String> reachedNines) {
        if (grid[r][c] == 9) {
            reachedNines.add(r + "," + c);
        }

        for (int d = 0; d < 4; d++) {
            int nr = r + dr[d];
            int nc = c + dc[d];

            if (nc >= 0 && nc < cols && nr >= 0 && nr < rows && grid[nr][nc] == grid[r][c] + 1) {
                dfsPart1(nr, nc, reachedNines);
            }
        }
    }

    static int SolvePart2() {
        int total = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] ==0 ){
                    total += dfsPart2(i, j);
                }
            }
        }
        return total;
    }

    static int dfsPart2(int r, int c) {
         if (grid[r][c] == 9) {
            return 1;
        }
        int paths = 0;
        for (int d = 0; d < 4; d++) {
            int nr = r + dr[d];
            int nc = c + dc[d];

            if (nc >= 0 && nc < cols && nr >= 0 && nr < rows && grid[nr][nc] == grid[r][c] + 1) {
                paths += dfsPart2(nr, nc);
            }
        }
        return paths;
    }
}

File: day10/day10.jl
dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

function read_input(filename)
    lines = readlines(filename)
    rows = length(lines)
    cols = length(lines[1])

    grid = zeros(Int, rows, cols)
    for i in 1:rows
        for j in 1:cols
            grid[i, j] = lines[i][j] - '0'
        end
    end
    return grid
end

function solve_part1(grid, rows, cols)
    total = 0
    for i in 1:rows
        for j in 1:cols
            if grid[i, j] == 0
                reached = Set{Tuple{Int,Int}}()
                dfsPart1(grid, rows, cols, i, j, reached)
                total += length(reached)
            end
        end
    end
    return total
end

function dfsPart1(grid, rows, cols, r, c, reached)
    if grid[r, c] == 9
        push!(reached, (r, c))
        return
    end

    for (dr, dc) in dirs
        nr = r + dr
        nc = c + dc

        if 1 <= nr <= rows && 1 <= nc <= cols && grid[nr, nc] == grid[r, c] + 1
            dfsPart1(grid, rows, cols, nr, nc, reached)
        end
    end
end

function solve_part2(grid, rows, cols)
    total = 0
    for i in 1:rows
        for j in 1:cols
            if grid[i, j] == 0
                total += dfsPart2(grid, rows, cols, i, j)
            end
        end
    end
    return total
end


function dfsPart2(grid, rows, cols, i, j)
    path = 0
    if grid[i, j] == 9
        return 1
    end

    for (dr, dc) in dirs
        nr = i + dr
        nc = j + dc

       if 1 <= nr <= rows && 1 <= nc <= cols && grid[nr, nc] == grid[i, j] + 1
            path += dfsPart2(grid, rows, cols, nr, nc)
        end
    end
    return path
end





grid = read_input("day10_inputs.txt")
rows, cols = size(grid)

println("Part 1 = ", solve_part1(grid, rows, cols))
println("Part 2 = ", solve_part2(grid, rows, cols))
